kvs/src/main.rs

```diff
@@ -216,27 +238,45 @@ where
                             })
                             .push_to(y_send),
                         hf.start_tee()
-                            .map(|msg| {
+                        .partition(
+                        |x| matches!(x, Message::Batch(_, _)),
+                            hf.start_tee().map(move |msg| {
                                 if let Message::Batch((id, epoch), batch) = msg {
                                     Some((id, epoch, batch))
                                 } else {
                                     unreachable!()
                                 }
                             })
                             .push_to(x_send),
+                            hf.start_tee().map(|msg| {
+                                if let Message::Get(k, sender) = msg {
+                                    Some((k, sender))
+                                } else {
+                                    unreachable!()
+                                }
+                            })
+                            .push_to(reads_send),
+                        )
                     ),
                 );
 
-                hf.add_subgraph("read_handler", reads_recv.flatten().map(|k| (k, ())).stream_join::<_, _, _, MaxRepr<V>, MaxRepr<V>>(
-                    x_recv.flatten().flat_map(|(id, epoch, batch)| {
-                        println!("got batch: {:?} {:?} {:?}", id, epoch, batch);
-                        // TODO(justin): this doesn't do the clocks. I
-                        // think we need to do some weird join against a
-                        // singleton that I don't really understand yet.
-                        batch.into_iter()
-                    })).pull_to_push().for_each(|x| {
-                        println!("read: {:?}", x)
-                    })
+                let my_id = id;
+                hf.add_subgraph("read_handler", reads_recv.flatten().map(|(k, ch)| (k, ch)).half_hash_join::<_, _, _, ClockedDataRepr<V>, ClockedUpdateRepr<V>>(
```

Comment:
Is rustfmt failing here? Can we space it out a bit more

Reply:
Yeah I'm not sure why rustfmt does not seem to be working here. I will have to investigate at some point.